experimental/torch_xla2/torch_xla2/ops/jaten.py

```diff
@@ -4560,3 +4560,100 @@ def _euclidean_direct(x1, x2):
+  n = LU_data.shape[-2]
+  m = LU_data.shape[-1]
+  dim = min(n,m)
+  #print("input: lu.shape ", LU_data.shape, LU_pivots.shape)
+  if unpack_data:
+    L = jnp.tril(LU_data, k=-1)        # Extract lower triangle
+    #print("lower:", L)
+    eye = jnp.eye(n, m, dtype=LU_data.dtype)
+    #print("eye:", eye)
+    L = L + eye                        # Add ones to the diagonal of L
+    #print("lower+eye:", L)
+    start_indices = jnp.zeros(len(LU_data.shape), dtype=int)
+    limit_indices = list(LU_data.shape)
+    limit_indices[-1] = dim
+    #print("indices: ", start_indices, limit_indices)
+    L = jax.lax.slice(L, start_indices, limit_indices) # Reshape lower triangle to match pivot
+
+    U = jnp.triu(LU_data)              # Extract upper triangle
+    start_indices = jnp.zeros(len(LU_data.shape), dtype=int)
+    limit_indices = list(LU_data.shape)
+    limit_indices[-2] = dim
+    U = jax.lax.slice(U, start_indices, limit_indices) # Reshape upper triangle to match pivot
+    #print("upper:",U)
+  else:
+    L = torch.empty(torch.Size([0]))
+    U = torch.empty(torch.Size([0]))
+
+  if unpack_pivots:
+    if True:
+      pivots = LU_pivots - 1           # pivots are offset by 1 in jax
+
+      pivot_size = pivots.shape[-1]
+      #? m = pivot_size  # In LU decomposition, k = min(m, n), and here we assume m <= n
+
+      # use an identity to create permutation matrix
+      # tile the 2D identity matrix in other dims
+      leading_dims = LU_data.shape[:-2]
+      tile_shape = (*leading_dims, 1, 1)
+      identity = jnp.eye(n, dtype=jnp.float32)
+      identity = jnp.expand_dims(identity, axis=tuple(range(len(leading_dims))))
+      P = jnp.tile(identity, tile_shape)
+      print("debug: start permutation matrix:", P)
+
+      # Apply the swaps iteratively
+      for i in range(pivot_size):
+        # Get the swap indices
+        row_idx = jnp.arange(n) #? row_idx = jnp.arange(m) or (pivot_size)
+        col_idx = pivots[..., i]
+        print("r c :", row_idx, col_idx, i)
+
+        #print("debug: (1, 1):", P[...,1, 1])
+        #print("debug: (row_idx, i):", P[...,row_idx, i])
+        #print("debug: (0, col_idx):", P[...,0, col_idx])
+        #print("debug: (row_idx, col_idx):", P[...,row_idx, col_idx])
+        #print("debug: (row_idx, i):", P[...,row_idx, i])
+        #return
+        # Swap the columns in the permutation matrix
+        pi = P[..., i, row_idx]
+        pp = jnp.zeros(pi.shape, dtype=pi.dtype)
+        for j in range(len(col_idx)):
+          pp_row = P[..., col_idx[j] , row_idx][j]
+          print("debug: pp_row:", j, pp_row)
+          pp = pp.at[j].set(pp_row)
+        print("debug: (i, row_idx):", pi)
+        print("debug: (row_idx, col_idx):", pp)
+        P = P.at[..., i, row_idx].set(pp)
+        print("debug: p1:", P)
+        P = P.at[..., col_idx, row_idx].set(pi)
+        print("debug: p2:", P)
+    else:
+      _pivots = LU_pivots - 1           # pivots are offset by 1 in jax
```

Comment:
this branch looks more correct according to this explanation:
```
Pivot Matrix Representation

PyTorch uses a compact representation for the permutation matrix. Instead of storing the full matrix, it stores a 1-indexed vector called pivots.

Each element pivots[i] indicates the row that the i-th row was swapped with during the LU decomposition process.
Importantly, it uses 1-indexing, meaning pivots[i] = j signifies that the i-th row was swapped with the (j-1)-th row.
Your Example: [2, 2] -> [[0, 1], [1, 0]]

Let's analyze your example step-by-step:

Input pivots = [2, 2]: This vector tells us:

In the 1st step, the 1st row was swapped with the (2-1) = 1st row (essentially, no swap).
In the 2nd step, the 2nd row was swapped with the (2-1) = 1st row.
Constructing the Permutation Matrix:  We start with an identity matrix and apply the swaps indicated by the pivots vector:

Initial Identity Matrix:

[[1, 0],
 [0, 1]] 
Step 1 (no swap): The matrix remains unchanged.

Step 2 (swap rows 2 and 1):

[[0, 1],
 [1, 0]]
```